### ISSUE TITLE
How to choose the validation dataset in ESC-50?

### ISSUE THREAD

**Issue by YazhouZhang0709** [karoldvl/paper-2015-esc-convnet#2]
Hi, @karoldvl . I just read your code and get a little confused about how the 5-fold cross-validation is done.
Could you give me some details about how you divide the dataset to 5 fold? 
I am new to this. Thank you!

**Comment by karolpiczak** [comment]
Every recording of the `ESC-50` dataset has a predefined fold. It's designated by the very first character in the filename of the recording. For instance, `ESC-50/101 - Dog/1-100032-A.ogg` gets into fold 1, and `3-144028-A.ogg` will go to fold 3.

In the end we have 5 folds [*1-5*], each fold will have exactly 8 recordings for each class.

When training the models we select and utilize 3 folds as training data, 1 fold for validation purposes/hyper-parameter tuning etc., and 1 for final testing (out-of-sample).

For instance we could use:
Folds `1`, `2`, `3` together as training.
Fold `4` as validation set.
Fold `5` as testing set.

We could also use those folds differently:
Training: `2`, `3`, `4`
Validation: `5`
Testing: `1`

In the end, this way we can evaluate the performance of the model on 5 independent out-of-sample/testing subsets of the whole dataset, and than aggregate those results. The results reported in the paper were achieved on the following setup:

**Fold** | Train | Validation | Test
----------|------|------------|-------
**1** | 2, 3, 4 | 5 | 1
**2** | 3, 4, 5 | 1 | 2
**3** | 1, 4, 5 | 2 | 3
**4** | 1, 2, 5 | 3 | 4
**5** | 1, 2, 3 | 4 | 5

The `Fold` column indicates how this particular setup is named in the [`results.csv`](https://github.com/karoldvl/paper-2015-esc-convnet/blob/master/Code/results.csv):

dataset | fold | method | accuracy
---------|------|----------|----------
ESC-50 | 1 | CNN - long/probability | 0.6475
ESC-50 | 2 | CNN - long/probability | 0.5700
ESC-50 | 3 | CNN - long/probability | 0.6150
ESC-50 | 4 | CNN - long/probability | 0.6825
ESC-50 | 5 | CNN - long/probability | 0.6275

The idea here is to get a less biased estimate of what the performance of the trained model would look like on new data with potentially different characteristics (yet still the same classes). It's not perfect, but it gives a good hint if we're overfitting.

Some caveats:
- It's possible to split the dataset into completely different folds (e.g.: random selection of recordings), but there are two major advantages of staying with those predefined ones. First thing, it's easier to compare the results you get. Second, I've restricted those parts which came from the same Freesound recording (different segments of the same original file) so that they are always in the same fold. This means that potential information leak between training/testing subsets should be limited (though there are still some possible issues with recording devices, preprocessing/bandwidth limiting).
- It's not that visible in the paper, but after exploring multiple variants on the validation set, you could also train your final model on training+validation data, and then evaluate on the test set. This has not been done in the paper, but should give you even better results than those reported. Especially, since these folds are quite small. Instead of discarding 20% of all data for validation purposes, we could enlarge the training set by a third in this way.

If you happen to need some further clarifications on this, ping me, I will try to address it.

**Comment by YazhouZhang0709** [comment]
Thanks so much!
